notebook/agentchat_sql.ipynb

```diff
@@ -0,0 +1,858 @@
+    "    ans = []\n",
+    "    for table in cursor.fetchall():\n",
+    "        ans.append(table[0])\n",
+    "    return ans\n",
+    "    \n",
+    "# Function to show the schema of a table\n",
+    "def show_table_schema(cursor, table_name):\n",
+    "    cursor.execute(f\"PRAGMA table_info({table_name});\")\n",
+    "    \n",
+    "    ans = f\"Here is the table information for Table: {table_name}\\n\"\n",
+    "    ans += \"-\" * 30 + \"\\n\"\n",
+    "    ans += \"Column ID | Column Name | Data Type | Not Null Constraint | Default Val | Primary Key | \\n\"\n",
+    "    ans += \"-\" * 30 + \"\\n\"\n",
+    "    for row in cursor.fetchall():\n",
+    "        ans += \" | \".join([str(x) for x in row]) + \"\\n\"\n",
+    "    ans += \"-\" * 30 + \"\\n\\n\\n\"\n",
+    "    return ans\n",
+    "\n",
+    "# Function to execute a query\n",
+    "def execute_query(cursor, query):\n",
+    "    cursor.execute(query)\n",
+    "    return cursor.fetchall()\n",
+    "\n",
+    "\n",
+    "def find_potential_joins(cursor):\n",
+    "    # Query to get foreign key relationships\n",
+    "    cursor.execute(\"SELECT tbl_name, sql FROM sqlite_master WHERE type='table'\")\n",
+    "    tables = cursor.fetchall()\n",
+    "\n",
+    "    joins = []\n",
+    "    for table, create_statement in tables:\n",
+    "        cursor.execute(f\"PRAGMA foreign_key_list({table})\")\n",
+    "        foreign_keys = cursor.fetchall()\n",
+    "        for fk in foreign_keys:\n",
+    "            # Each item in foreign_keys is a tuple like:\n",
+    "            # (id, seq, table, from, to, on_update, on_delete, match)\n",
+    "            from_table = table\n",
+    "            from_column = fk[3]\n",
+    "            to_table = fk[2]\n",
+    "            to_column = fk[4]\n",
+    "            join_info = f\"{from_table}.{from_column} -> {to_table}.{to_column}\"\n",
+    "            joins.append(join_info)\n",
+    "\n",
+    "    return joins\n",
+    "\n",
+    "def column_names(cursor, table_name):\n",
+    "    \"\"\"\n",
+    "    Retrieves the column names for a specified table.\n",
+    "    \n",
+    "    Args:\n",
+    "    cursor (sqlite3.Cursor): The cursor object connected to the database.\n",
+    "    table_name (str): The name of the table for which to retrieve column names.\n",
+    "\n",
+    "    Returns:\n",
+    "    list: A list of column names for the specified table.\n",
+    "    \"\"\"\n",
+    "    query = f\"PRAGMA table_info({table_name});\"\n",
+    "    cursor.execute(query)\n",
+    "    # Fetches all rows from the query result, extracts the second column which contains column names\n",
+    "    return [row[1] for row in cursor.fetchall()]\n",
+    "\n",
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "id": "0179745d",
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "def table_prompt(cursor, question=\"\", model=\"gpt-4\", max_tokens=10e9):\n",
```

Comment:
max_tokens=10e9 - one billion tokens is a bit too much) I would change it to 4096 for to be a more realistic/conservative value compatible with most of the models